torch/jit/annotations.py

```diff
@@ -335,6 +335,8 @@ def try_ann_to_type(ann, loc):
     # Maybe resolve a NamedTuple to a Tuple Type
     def fake_rcb(key):
         return None
+    # If ann is a NamedTuple, check that it meet requirements for JIT compilation.
+    torch._jit_internal.check_can_compile_named_tuple(ann, loc)
```

Comment:
Should this be done in C++ instead?

Reply:
Good point, let me move this to C++ and call it only when we are sure we are dealing with a `NamedTuple`:

```
    if (isNamedTupleClass(obj)) {
      // Maybe here or inside registerNamedTuple
      return registerNamedTuple(obj, loc);
    }
```